File: JavaScript/Trees/Binary_Tree/BinaryTree.js
/*******************************************************
*  LinkedList.js
*  Created by Stephen Hall on 6/29/16.
*  Copyright (c) 2016 Stephen Hall. All rights reserved.
*  A Binary Tree implementation in JavaScript
********************************************************/

/** 
 * Node Object declaration
 * 
 * @param  generic data
 *
 * @return none
 * @throws none
 **/
function Node(data){
    this.parent = null;
    this.left = null;
    this.right = null;
    this.data = data; 
};

/** 
 * Binary Tree Object declaration
 *          
 * @param  none
 *
 * @return none
 * @throws none
 **/
function BinaryTree(){
    this.root = null;
    
    /** 
     * Node compare function
     *      Compares nodes data to data for equality
     * 
     * @param  generic data, node to compare
     *
     * @return int to represent greater then, less then, equal to
     * @throws none
     **/
    this.compare = function(data, node){
        if(node == null || data == null)
            return null;
        
        if(node.data > data)
            return 1;
        
        if(node.data < data)
            return -1;
        
        if(node.data == data)
            return 0;
        
        return null;
    };
    
    /** 
     * Insert Helper method
     *      Inserts node into the tree
     *
     * @param  generic data, compare node
     *
     * @return Inserted Node
     * @throws none
     **/
    this.insertHelper = function(data, node){
        var cmp = this.compare(data, node);
        if(cmp != null) {
            if(cmp == 1){
                if(!node.right){
                    newNode = new Node(data);
                    node.right = newNode;
                    newNode.parent = node;
                    return newNode;
                }
                else
                    return this.insertHelper(data, node.right);
            }
            if(cmp <= 0){
                if(!node.left){
                    newNode = new Node(data);
                    node.left = newNode;
                    newNode.parent = node;
                    return newNode;
                }
                else
                    return this.insertHelper(data, node.left);
            }
        }
        else
            return null;
    };
    
    /** 
     * Remove helper function
     *          Removes node from the tree
     *
     * @param  generic data, compare node
     *
     * @return removed node
     * @throws none
     **/
    this.removeHelper = function(data, node){
        var cmp = this.compare(data, node);
        if(cmp != null){
            if(cmp == 1)
                return this.removeHelper(data, node.right);
            if(cmp == -1)
                return this.removeHelper(data, node.left);
            if(cmp == 0){
                //has no children

                var tempNode = null;

                if(!node.left && !node.right) {
                    node = node.parent;
                }
                
                //has one child
                if(node.parent) { 
                	tempNode = node.parent;
                }
                else { // this is the root node
                	if(node.right) {
                		tempNode = node.right;
                		tempNode.parent = null;
                		this.root.right = null;
                		this.root = tempNode;
                		return node;
                	}
                	else {
            			tempNode = node.left;
            			tempNode.parent = null;
                		this.root.left = null;
                		this.root = tempNode;
                		return node;
                	}
                }
                
                if(tempNode.left === node) {
        			if(node.left) {
        				tempNode.left = node.left;
        				node.left.parent = tempNode;
        				return node;
        			}
        			else if(node.right) {
        				tempNode.right = node.right;
        				node.right.parent = tempNode;
        				return node;
        			}
            	}
	            if(tempNode.right === node) {
        			if(node.left) {
        				tempNode.left = node.left;
        				node.left.parent = tempNode;
        				return node;
        			}
        			else if(node.right) {
        				tempNode.right = node.right;
        				node.right.parent = tempNode;
        				return node;
        			}
            	}

                else if (node.left && node.right) { // test for 2 children
                	tempNode = node.right; // find min value of right child tree to replace the node
                	while(tempNode.left)
        				tempNode = tempNode.left;
        			var temp = node.data;
                	node.data = tempNode.data; // replace the node with the tempNode
                	tempNode.data = temp;
                	tempNode.parent.left = null;
                	tempNode.parent = null;
                	return tempNode;
                }
                
            }

        }
        else
            return null;
    };
    
}

/** 
 * Insert prototype function
 *          Inserts node into the tree
 *
 * @param  generic data
 *
 * @return Inserted node
 * @throws none
 **/
BinaryTree.prototype.insert = function(data){
    if(!this.root){
       return this.root = new Node(data);
    }
    console.log("added " + data);
    return this.insertHelper(data, this.root);

};

/** 
 * Remove prototype function
 *          Removes node from the tree
 *
 * @param  generic data 
 *
 * @return removed node
 * @throws none
 **/
BinaryTree.prototype.remove = function(data) {
    if(!this.root)
        return null;
    return this.removeHelper(data, this.root);

};

/** 
 * Get Minimum value prototype function
 *          finds the smallest value in the tree
 *
 * @param  none
 *
 * @return smallest node
 * @throws none
 **/
BinaryTree.prototype.getMin = function(){
    if(!this.root)
        return null;
    
    node = this.root;
    
    while(node.left)
        node = node.left;
    
    return node;
};

/** 
 * Get Maximum value prototype function
 *          finds the largest value in the tree
 *
 * @param  none
 *
 * @return largest node
 * @throws none
 **/
BinaryTree.prototype.getMax = function(){
    if(!this.root)
        return null;
    
    node = this.root;
    
    while(node.right)
        node = node.right;
    
    return node;
};


/** 
 * Preorder traversal prototype function
 *          preforms recursive preorder traversal on the tree
 *
 * @param  start node
 *
 * @return none
 * @throws none
 **/
BinaryTree.prototype.preOrderTraversal = function(node){
    if(node !== null){
        console.log(node.data);
        this.preOrderTraversal(node.left);
        this.preOrderTraversal(node.right);
    }
};

/** 
 * Post traversal prototype function
 *          preforms recursive Postorder traversal on the tree
 *
 * @param  start node
 *
 * @return none
 * @throws none
 **/
BinaryTree.prototype.postOrderTraversal = function(node){
    if(node != null){
        this.postOrderTraversal(node.left);
        this.postOrderTraversal(node.right);
        console.log(node.data);
    }
};

/** 
 * Inorder traversal prototype function
 *          preforms recursive inorder traversal on the tree
 *
 * @param  start node
 *
 * @return none
 * @throws none
 **/
BinaryTree.prototype.inOrderTraversal = function(node){
    if(node != null){
        this.inOrderTraversal(node.left);
        console.log(node.data);
        this.inOrderTraversal(node.right);
    }
};

/** 
 * Depth Frist Search prototype function
 *          preforms Depth First Search on the tree
 *
 * @param  start node
 *
 * @return none
 * @throws none
 **/
BinaryTree.prototype.depthFirstSearch = function(node){
		if(node == null)
            return;

        var stack = [];
        stack.push(node);
        
		while (stack.length > 0) {
			console.log(node = stack.pop());
			if(node.right!=null) 
                stack.push(node.right);
			if(node.left!=null) 
                stack.push(node.left);			
		}
};
	
/** 
 * Breadth First Search prototype function
 *          preforms Breadth First Search on the tree
 *
 * @param  start node
 *
 * @return none
 * @throws none
 **/
BinaryTree.prototype.breadthFirstSearch = function(node){
    if(node == null)
        return;

    var queue = [];
    queue.push(node);

    while (queue.length > 0) {
        console.log((node = queue.shift()));
        if(node.left != null) 
            queue.push(node.left);			
        if(node.right != null) 
            queue.push(node.right);

    }
};

File: C#/Trees/Binary_Tree/BinaryTree.cs
/*******************************************************
 *  BinaryTree.cs
 *  Created by Stephen Hall on 9/25/17.
 *  Copyright (c) 2016 Stephen Hall. All rights reserved.
 *  A Binary Tree implementation in C#
 ********************************************************/
using System;
using System.Collections.Generic;

namespace DataStructures.Trees.BinaryTree
{
    /// <summary>
    /// Binary Tree Class
    /// </summary>
    /// <typeparam name="T">Generic Type</typeparam>
    public class BinaryTree<T> where T : IComparable
    {
        /// <summary>
        /// Node class for the binary tree
        /// </summary>
        public class Node
        {
            /// <summary>
            /// Public Properties for  the Node class
            /// </summary>
            public T Data { set; get; }
            public Node Left { set; get; }
            public Node Right { set; get; }
            public Node Parent { set; get; }

            /// <summary>
            /// Node class constructor             
            /// </summary>
            /// <param name="data">Data to be held in the Node</param>
            public Node(T data)
            {
                Left = Right = null;
                Data = data;
            }
        }

        /// <summary>
        /// Private Binary Tree members
        /// </summary>
        private Node _root;

        /// <summary>
        /// Binary tree class constructor
        /// </summary>
        public BinaryTree() => _root = null;

        /// <summary>
        /// Compares the data in given node to the give data for equality
        /// </summary>
        /// <param name="data">data to compare</param>
        /// <param name="node">Node to compare to</param>
        /// <returns>1 if greater then, 0 if equal to, -1 if less then</returns>
        private int Compare(T data, Node node) => node.Data.CompareTo(data);

        /// <summary>
        /// Inserts a new node into the tree
        /// </summary>
        /// <param name="data">data to insert into the tree</param>
        /// <returns>Node inserted into the tree</returns>
        public Node Insert(T data)
        {
            return _root == null ? _root = new Node(data) : InsertHelper(data, _root);
        }
        
        /// <summary>
        /// Recursive helper function to inset a new node into the tree
        /// </summary>
        /// <param name="data">Data to insert into the tree</param>
        /// <param name="node">urrent node in the tree</param>
        /// <returns>Node inserted into the tree</returns>
        private Node InsertHelper(T data, Node node)
        {
            try
            {
                int cmp = Compare(data, node);
                switch (cmp)
                {
                    case 1:
                        if (node.Right == null)
                        {
                            Node newNode = new Node(data);
                            node.Right = newNode;
                            newNode.Parent = node;
                            return newNode;
                        }
                        else
                            return InsertHelper(data, node.Right);

                    default:
                        if (node.Left == null)
                        {
                            Node newNode = new Node(data);
                            node.Left = newNode;
                            newNode.Parent = node;
                            return newNode;
                        }
                        else
                            return InsertHelper(data, node.Left);
                }
            }
            catch (Exception)
            {
                return null;
            }
        }

        /// <summary>
        /// Removes a Node from the tree
        /// </summary>
        /// <param name="data">Data to remove from the tree</param>
        /// <returns>Node removed from the tree</returns>
        public Node Remove(T data) => _root != null ? null : RemoveHelper(data, _root);

        /// <summary>
        /// Recursive helper function to remove a node from the tree
        /// </summary>
        /// <param name="data">Data to remove</param>
        /// <param name="node">urrent node</param>
        /// <returns>Node removed from the tree</returns>
        private Node RemoveHelper(T data, Node node)
        {
            try
            {
                int cmp = Compare(data, node);
                if (cmp == 1)
                    return RemoveHelper(data, node.Right);
                if (cmp == -1)
                    return RemoveHelper(data, node.Left);
                if (cmp == 0)
                {
                    //has no children
                    Node tempNode;

                    if (node.Left == null && node.Right == null)
                        node = node.Parent;
                    
                    //has one child
                    if (node.Parent != null)
                        tempNode = node.Parent;
                    else
                    { // this is the root node
                        if (node.Right != null)
                        {
                            tempNode = node.Right;
                            tempNode.Parent = null;
                            _root.Right = null;
                            _root = tempNode;
                            return node;
                        }
                        tempNode = node.Left;
                        if (tempNode != null)
                        {
                            tempNode.Parent = null;
                            _root.Left = null;
                            _root = tempNode;
                        }
                        return node;
                    }
                    if (tempNode.Left == node)
                    {
                        if (node.Left != null)
                        {
                            tempNode.Left = node.Left;
                            node.Left.Parent = tempNode;
                            return node;
                        }
                        if (node.Right != null)
                        {
                            tempNode.Right = node.Right;
                            node.Right.Parent = tempNode;
                            return node;
                        }
                    }
                    if (tempNode.Right == node)
                    {
                        if (node.Left != null)
                        {
                            tempNode.Left = node.Left;
                            node.Left.Parent = tempNode;
                            return node;
                        }
                        if (node.Right != null)
                        {
                            tempNode.Right = node.Right;
                            node.Right.Parent = tempNode;
                            return node;
                        }
                    }
                    else if (node.Left != null && node.Right != null)
                    { // test for 2 children
                        tempNode = node.Right; // find min value of right child tree to replace the node
                        while (tempNode.Left != null)
                            tempNode = tempNode.Left;
                        T temp = node.Data;
                        node.Data = tempNode.Data; // replace the node with the tempNode
                        tempNode.Data = temp;
                        tempNode.Parent.Left = null;
                        tempNode.Parent = null;
                        return tempNode;
                    }
                }
            }
            catch (Exception) { return null; }
            return null;
        }
        
        /// <summary>
        /// Returns the smallest node in the tree
        /// </summary>
        /// <returns>Smallest Node int he tree</returns>
        public Node GetMin()
        {
            if (_root == null)
                return null;

            Node node = _root;

            while (node.Left != null)
                node = node.Left;

            return node;
        }
        
        /// <summary>
        /// Return the largest node in the tree
        /// </summary>
        /// <returns>argest Node in the tree</returns>
        public Node GetMax()
        {
            if (_root == null)
                return null;

            Node node = _root;

            while (node.Right != null)
                node = node.Right;

            return node;
        }
       
        /// <summary>
        /// Prints out the tree using Pre Order Traversal
        /// </summary>
        /// <param name="node">Node to start the Pre Order Traversal at</param>
        public void PreOrederTraversal(Node node)
        {
            if (node != null)
            {
                Console.WriteLine(node.Data);
                PreOrederTraversal(node.Left);
                PreOrederTraversal(node.Right);
            }
        }
        
        /// <summary>
        /// Prints out the Tree using Post Order Traversal
        /// </summary>
        /// <param name="node">Node to start the Post Order Traversal at</param>
        public void PostPrderTraversal(Node node)
        {
            if (node != null)
            {
                PostPrderTraversal(node.Left);
                PostPrderTraversal(node.Right);
                Console.WriteLine(node.Data);
            }
        }
        
        /// <summary>
        /// Pints out the tree using in order Traversal
        /// </summary>
        /// <param name="node">Node to start the In Order Traversal at</param>
        public void InOrderedTraversal(Node node)
        {
            if (node != null)
            {
                InOrderedTraversal(node.Left);
                Console.WriteLine(node.Data);
                InOrderedTraversal(node.Right);
            }
        }
        
        /// <summary>
        /// Prints out the tree using Depth First Search
        /// </summary>
        /// <param name="node">Node to start the Depth First Search at</param>
        public void DepthFirstSearch(Node node)
        {
            if (node == null)
                return;

            Stack<Node> stack = new Stack<Node>();
            stack.Push(node);

            while (stack.Count > 0)
            {
                Console.WriteLine(node = stack.Pop());
                if (node.Right != null)
                    stack.Push(node.Right);
                if (node.Left != null)
                    stack.Push(node.Left);
            }
        }
        
        /// <summary>
        /// Prints out the tree using breadth first search
        /// </summary>
        /// <param name="node">Node to start Breadth First Search at</param>
        public void BreadthFirstSearch(Node node)
        {
            if (node == null)
                return;

            Queue<Node> queue = new Queue<Node>();
            queue.Enqueue(node);

            while (queue.Count > 0)
            {
                Console.WriteLine(node = queue.Dequeue());
                if (node.Left != null)
                    queue.Enqueue(node.Left);
                if (node.Right != null)
                    queue.Enqueue(node.Right);
            }
        }
    }

}
